File: 203.remove-linked-list-elements.cpp
/*
 * @lc app=leetcode id=203 lang=cpp
 *
 * [203] Remove Linked List Elements
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
 public:
  ListNode* removeElements(ListNode* head, int val) {
    while (head) {
      if (head->val != val) {
        break;
      }
      head = head->next;
    }
    ListNode* node = head;
    while (node) {
      ListNode* next = node->next;
      while (next) {
        if (next->val == val) {
          node->next = next->next;
          next = node->next;
        } else {
          break;
        }
      }
      node = node->next;
    }
    return head;
  }
};
// @lc code=end

File: 203.remove-linked-list-elements.rs
/*
 * @lc app=leetcode id=203 lang=rust
 *
 * [203] Remove Linked List Elements
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn remove_elements(mut head: Option<Box<ListNode>>, val: i32) -> Option<Box<ListNode>> {
        let mut list = None;
        let mut tail = &mut list;

        while let Some(mut node) = head.take() {
            head = node.next.take();

            if node.val != val {
                tail = &mut tail.insert(node).next;
            }
        }
        list
    }
}
// @lc code=end

#[test]
#[allow(unused_assignments)]
fn test1() {
    assert_eq!(
        Solution::remove_elements(list_node![1, 2, 6, 3, 4, 5, 6], 6),
        list_node![1, 2, 3, 4, 5]
    );
}

#[test]
fn test2() {
    assert_eq!(Solution::remove_elements(list_node![], 1), list_node![]);
}

#[test]
#[allow(unused_assignments)]
fn test3() {
    assert_eq!(
        Solution::remove_elements(list_node![7, 7, 7, 7], 7),
        list_node![]
    );
}
